akka-persistence-query/src/main/scala/akka/persistence/query/typed/internal/EventsBySliceFirehose.scala

```diff
@@ -0,0 +1,630 @@
+import akka.annotation.InternalApi
+import akka.event.Logging
+import akka.event.LoggingAdapter
+import akka.persistence.query.Offset
+import akka.persistence.query.PersistenceQuery
+import akka.persistence.query.TimestampOffset
+import akka.persistence.query.typed.EventEnvelope
+import akka.persistence.query.typed.scaladsl.EventsBySliceQuery
+import akka.stream.Attributes
+import akka.stream.FanInShape2
+import akka.stream.FlowShape
+import akka.stream.Inlet
+import akka.stream.KillSwitch
+import akka.stream.KillSwitches
+import akka.stream.Outlet
+import akka.stream.scaladsl.BroadcastHub
+import akka.stream.scaladsl.GraphDSL
+import akka.stream.scaladsl.Keep
+import akka.stream.scaladsl.Source
+import akka.stream.stage.GraphStage
+import akka.stream.stage.GraphStageLogic
+import akka.stream.stage.InHandler
+import akka.stream.stage.OutHandler
+import akka.stream.stage.StageLogging
+import akka.util.JavaDurationConverters._
+import akka.util.OptionVal
+import akka.util.unused
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object EventsBySliceFirehose
+    extends ExtensionId[EventsBySliceFirehose]
+    with ExtensionIdProvider {
+
+  override def get(system: ActorSystem): EventsBySliceFirehose = super.get(system)
+
+  override def get(system: ClassicActorSystemProvider): EventsBySliceFirehose = super.get(system)
+
+  override def lookup = EventsBySliceFirehose
+
+  override def createExtension(system: ExtendedActorSystem): EventsBySliceFirehose =
+    new EventsBySliceFirehose(system)
+
+  object Settings {
+    def apply(system: ActorSystem, pluginId: String): Settings =
+      apply(system.settings.config.getConfig(pluginId))
+
+    def apply(config: Config): Settings =
+      Settings(
+        delegateQueryPluginId = config.getString("delegate-query-plugin-id"),
+        broadcastBufferSize = config.getInt("broadcast-buffer-size"),
+        firehoseLingerTimeout = config.getDuration("firehose-linger-timeout").asScala,
+        catchupOverlap = config.getDuration("catchup-overlap"),
+        slowConsumerReaperInterval = config.getDuration("slow-consumer-reaper-interval").asScala,
+        slowConsumerBehindThreshold = config.getDuration("slow-consumer-behind-threshold"),
+        abortSlowConsumerAfter = config.getDuration("abort-slow-consumer-after"))
+  }
+
+  final case class Settings(
+      delegateQueryPluginId: String,
+      broadcastBufferSize: Int,
+      firehoseLingerTimeout: FiniteDuration,
+      catchupOverlap: JDuration,
+      slowConsumerReaperInterval: FiniteDuration,
+      slowConsumerBehindThreshold: JDuration,
+      abortSlowConsumerAfter: JDuration) {
+    require(
+      delegateQueryPluginId != null && delegateQueryPluginId.nonEmpty,
+      "Configuration of delegate-query-plugin-id must defined.")
+  }
+
+  final class SlowConsumerException(message: String) extends RuntimeException(message) with NoStackTrace
+
+  final case class FirehoseKey(pluginId: String, entityType: String, sliceRange: Range)
```

Comment:
I intend to create a follow up ticket about colocating shards. So when several projections are started with ShardedDaemonProcess they would try to allocate slice ranges to the same nodes, so that firehose for a slice range is not started in more places than necessary.